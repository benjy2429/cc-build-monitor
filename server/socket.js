import IO from 'socket.io';
import resolve from './resolver';
import mockFetch from './mock-fetch';

const POLL_RATE = parseInt(process.env.POLL_RATE || '30000', 10);

const getData = (config = process.env) => {
  if (config.USE_MOCKS === 'true') {
    return resolve(mockFetch);
  }
  return resolve();
};

export default (server, socketLib = IO) => {
  const io = socketLib(server);

  io.on('connection', async (socket) => {
    socket.emit('monitorData', await getData());
  });

  setInterval(
    async () => {
      const connections = Object.keys(io.sockets.connected).length;
      if (connections) {
        io.sockets.emit('monitorData', await getData());
      }
    },
    POLL_RATE,
  );
};
