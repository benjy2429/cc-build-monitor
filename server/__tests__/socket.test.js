import socket from '../socket';
import resolve from '../resolver';

describe('Socket', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    resolve.default = jest.fn().mockReturnValue({ projects: [], builds: [] });
  });

  it('initialises the socket library with the server', () => {
    const mockLib = jest.fn().mockReturnValue({ on: () => {} });

    socket('some-server', mockLib);
    expect(mockLib.mock.calls.length).toBe(1);
    expect(mockLib.mock.calls[0][0]).toBe('some-server');
  });

  it('emits data when a new client connects', async () => {
    let onConn;
    const mockSocket = { emit: jest.fn() };
    const mockLib = {
      on: (event, cb) => { onConn = cb; },
    };

    socket(null, () => mockLib);
    await onConn(mockSocket);
    expect(mockSocket.emit.mock.calls.length).toBe(1);
    expect(mockSocket.emit.mock.calls[0][0]).toBe('monitorData');
    expect(JSON.parse(mockSocket.emit.mock.calls[0][1])).toMatchSnapshot();
  });

  it('emits data on a timer', async () => {
    global.setInterval = jest.fn();
    const mockEmit = jest.fn();
    const mockLib = {
      on: () => {},
      sockets: {
        connected: [1],
        emit: mockEmit,
      },
    };

    socket(null, () => mockLib);
    expect(setInterval.mock.calls.length).toBe(1);

    await setInterval.mock.calls[0][0]();
    expect(mockEmit.mock.calls.length).toBe(1);
    expect(mockEmit.mock.calls[0][0]).toBe('monitorData');
    expect(JSON.parse(mockEmit.mock.calls[0][1])).toMatchSnapshot();
  });

  it('does not emit data if no clients are connected', async () => {
    global.setInterval = jest.fn();
    const mockEmit = jest.fn();
    const mockLib = {
      on: () => {},
      sockets: {
        connected: [],
        emit: mockEmit,
      },
    };

    socket(null, () => mockLib);
    await setInterval.mock.calls[0][0]();
    expect(mockEmit.mock.calls.length).toBe(0);
  });
});
