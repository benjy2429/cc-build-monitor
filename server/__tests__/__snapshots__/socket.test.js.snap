// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Socket emits data on a timer 1`] = `
Object {
  "projects": Array [],
}
`;

exports[`Socket emits data when a new client connects 1`] = `
Object {
  "projects": Array [],
}
`;
